broadinstitute/single_cell_portal_core
# Adding visualization facets API endpoint, indexed cluster cell arrays (SCP-5214)

#### BACKGROUND
In order to support some of the planned features to allow users to visualize intersections & subsets of data (SCP-5178), we need to have data available to quickly find annotation memberships for populations of cells in a scatter plot.  This will allow us to build a "faceted search" feature within a plot that will let users progressively filter data based off of facet (i.e. annotation names) and filter (i.e. group-based annotation values) values.  Adding an API endpoint that will serve this data on demand in a performant manner is one of the first requirements.

#### CHANGES
This update adds that endpoint at `/api/v1/studies/:accession/annotations/facets` (in `Api::V1::Visualization::AnnotationsController#facets`).  This endpoint will compute integer-based assignments for the requested cluster/annotations.  A functional spec is available [here](https://docs.google.com/document/d/1OHJBF_6DnKCGgf4jwuZjq2iZJF5kDEl5eG23PLGVOR4/edit#heading=h.gcdzl0yfknth) with more detail on request parameters/responses.  In the response, the `cells` array shows where each cell from the cluster (outer array) maps to annotation values (inner arrays) as defined in the `facets` array, pulling those values from the `groups` key.

One of the major requirements for this was performance - it has to be able to return this data quickly (1-2s, ideally).  The initial implementation was woefully inadequate (25-30s for 50K cells, 2 annotations).  To address this problem, all clusters will now index their cell name arrays against the "all cells" array from the study metadata file.  By doing this, we pre-calculate the order in which a given cluster's cells appear in any study-based annotation (cluster-based annotations are not affected by this problem).  This now allows us to skip loading the entire hash of study based annotations, and instead _only_ retrieve the array of annotation labels.  Then, using the indexed cluster cell array, we can pull out the correct values for each cell only using the array index values.

Toy example:
```
metadata_cells = ['A', 'B', 'C', 'D', 'E']
metadata_labels = ['foo', 'bar', 'bing', 'baz', 'blah'] 
annotation_hash = { A: 'foo', B: 'bar', C: 'bing', D: 'baz', E: 'blah' }
cluster_cells = ['C', 'E', 'A', 'D', 'B']
indexed_cluster_cells = [2, 4, 0, 3, 1]

$ cluster_cells.first
=> 'C'
$ annotation_hash['C'] 
=> 'bing'
$ indexed_cluster_cells.first 
=> 2
$ metadata_labels[2] 
=> 'bing'
```

Benchmarking with production data has shown this approach to be 3-7x faster than using the entire hash to load the data (render/plotting times would be unaffected).  This performance boost holds at all resolutions, but is not appreciable at under 100K cells, and therefore is only applied to full resolution data (also since subsampling does not always sample the same cells across annotations, making backfill infeasible).  This update applies this technique to all visualizations that use full resolution study-based data.  Additionally, a migration has been added to backfill the necessary data, and hooks have been added to `IngestJob` to generate the arrays whenever cluster/metadata files are ingested (along with data cleanup calls in `DeleteQueueJob`).

The hope is that this performance enhancement will eventually remove the need to subsample data, which would be a major complexity/cost savings for the product, and a gain in performance/usability for all users.

#### MANUAL TESTING
If you plan to use the Swagger doc to test, you will need to make the following edit to `app/views/api/v1/api_docs/swagger_ui.html.erb` at the bottom of the page where Swagger is initialized:

```
 const ui = SwaggerUIBundle({
  syntaxHighlight: false,
  ...
```
This will prevent Swagger from crashing on the large array response.  Additionally, disable caching by running this in the terminal from project root:
```
bin/rails dev:cache
=> Development mode is no longer being cached.
```
Alternatively, you can use a client like [Postman](https://www.postman.com/).
1. Start all services, including `Delayed::Job` and then run the migration with `bin/rails db:migrate`
2. Load the `Human milk - differential expression` study and copy the study accession
3. Using Swagger, load `https://localhost:3000/single_cell/api/v1#/Visualization/study_annotation_facets_path`, or point Postman at `https://localhost:3000/single_cell/api/v1/studies/[:accession]/annotations/facets`, adding in the correct `:accession` value
4. Use the following request parameters:
    * accession: from above
    * annotations: `General_Celltype--group--study,disease__ontology_label--group--study`
    * cluster: 'All Cells UMAP'
5. Execute the request, and confirm the response comes back quickly (~3s, depending on your internet connection)
6. There should be a long array of nested two-element arrays for `cells`, and `facets` at the bottom should list both annotations and their values.
7. In a separate tab, load the same [study on production](https://singlecell.broadinstitute.org/single_cell/study/SCP1671/cellular-and-transcriptional-diversity-over-the-course-of-human-lactation)
8. Run the same visualization requests (normal & expression-based) in both tabs, confirming that you see the same plots (colors will be different in the scatter plots on production)
